integrated_channels/blackboard/client.py

```diff
@@ -40,15 +45,235 @@ def __init__(self, enterprise_configuration):
+
+        Example payload:
+        ---------------
+            '{
+                courseID: course-edx+555+3T2020,
+                score: 0.85,
+                completedTimestamp: 1602265162589,
+            }'
+
+        """
+        self._create_session()
+        learner_data = json.loads(payload)
+        external_id = learner_data.get('courseID')
+
+        course_id = self._resolve_blackboard_course_id(external_id)
+
+        # Sanity check for course id
+        if not course_id:
+            raise ClientError(
+                'Could not find course:{} on Blackboard'.format(external_id),
+                HTTPStatus.NOT_FOUND.value
+            )
+
+        blackboard_user_id = self._get_bb_user_id_from_enrollments(user_id, course_id)
+        grade_column_id = self._get_or_create_integrated_grade_column(course_id)
+
+        grade = learner_data.get('grade') * 100
+        grade_percent = {'score': grade}
+        response = self._patch(
+            self.generate_post_users_grade_url(course_id, grade_column_id, blackboard_user_id),
+            grade_percent
+        )
+
+        if response.json().get('score') != grade:
+            raise ClientError(
+                'Failed to post new grade for user={} enrolled in course={}'.format(user_id, course_id),
+                HTTPStatus.INTERNAL_SERVER_ERROR.value
+            )
+
+        success_body = 'Successfully posted grade of {grade} to course:{course_id} for user:{user_email}.'.format(
+            grade=grade,
+            course_id=external_id,
+            user_email=user_id,
+        )
+        return response.status_code, success_body
+
+    def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-argument
+        """TODO: course completion deletion is currently not easily supported"""
+
+    @staticmethod
+    def _raise_for_external_id(channel_metadata_item):
+        """
+        Raise error if external_id invalid or not found
+        """
+        if 'externalId' not in channel_metadata_item:
+            raise ClientError("No externalId found in metadata, please check json data format", 400)
+
+    @staticmethod
+    def _raise_for_course_id(course_id, external_id):
+        """
+        Raise error if course_id invalid
+        """
+        if not course_id:
+            raise ClientError(
+                'Could not find course:{} on Blackboard'.format(external_id),
+                HTTPStatus.NOT_FOUND.value
+            )
+
+    def _resolve_blackboard_course_id(self, external_id):
+        """
+        Extract course id from blackboard, given it's externalId
+        """
+        params = {'externalId': external_id}
+        courses_responses = self._get(self.generate_courses_url(), params).json()
+        course_response = courses_responses.get('results')
+
+        for course in course_response:
+            if course.get('externalId') == external_id:
```

Comment:
if you're searching by externalId on Blackboards side, will this ever not be true?

Reply:
Besides a 404 that is.

Reply:
good point, more precisely really we could even just get the first item that is returned and use it without checking. I actually need to test this manually once, I __think__ it will work but I really have to be sure.. be back in a bit with this answer

Reply:
My edge case questions would be:
- what's the response with no courses present (will we even have to worry about 404's with the error handling?)
- what's the  response with other courses but not the one we're looking for (will we even have to worry about 404's with the error handling?)
- what's the successful response body when we find just our course and just our course exists (I've tested this, it's not really an edge case)
- what's the successful response body when there are multiple courses, including our's?

If those are consistent, then I think we could just go without checking!

Reply:
yeah I would think that for if any reason we can't find a course_id here we can bail out safely

Reply:
and tested this code just now works!